types/src/checksummed_hex.rs

```diff
@@ -0,0 +1,258 @@
+//! Checksummed hex encoding following an [EIP-55][1]-like scheme.
+//!
+//! [1]: https://eips.ethereum.org/EIPS/eip-55
+
+use alloc::{string::String, vec::Vec};
+use core::ops::RangeInclusive;
+
+use base16;
+use blake2::{Blake2b, Digest};
+
+/// The number of input bytes, at or below which [`encode`] will checksum-encode the output.
+pub const SMALL_BYTES_COUNT: usize = 75;
+
+const HEX_CHARS: [char; 22] = [
+    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f', 'A', 'B', 'C',
+    'D', 'E', 'F',
+];
+
+/// Takes a slice of bytes and breaks it up into a vector of *nibbles* (ie, 4-bit values)
+/// represented as `u8`s.
+fn bytes_to_nibbles(input: &(impl AsRef<[u8]> + ?Sized)) -> impl Iterator<Item = u8> + '_ {
+    input
+        .as_ref()
+        .iter()
+        .flat_map(move |byte| [4, 0].iter().map(move |offset| (byte >> offset) & 0x0f))
+}
+
+/// Takes a slice of bytes and outputs an infinite cyclic stream of bits for those bytes.
+fn bytes_to_bits_cycle(bytes: Vec<u8>) -> impl Iterator<Item = bool> {
+    bytes
+        .into_iter()
+        .cycle()
+        .flat_map(move |byte| (0..8usize).map(move |offset| ((byte >> offset) & 0x01) == 0x01))
+}
+
+/// Computes a Blake2b hash.
+fn blake2b_hash(data: impl AsRef<[u8]>) -> Vec<u8> {
+    let mut hasher = Blake2b::new();
+    hasher.update(data);
+    hasher.finalize().to_vec()
+}
+
+/// If `input` is not greater than [`SMALL_BYTES_COUNT`], returns the bytes encoded as hexadecimal
+/// with mixed-case based checksums following a scheme similar to [EIP-55][1].  If `input` is
+/// greater than `SMALL_BYTES_COUNT`, no mixed-case checksumming is applied, and lowercase hex is
+/// returned.
+///
+/// Key differences:
+///   - Works on any length of data up to `SMALL_BYTES_COUNT`, not just 20-byte addresses
+///   - Uses Blake2b hashes rather than Keccak
+///   - Uses hash bits rather than nibbles
+///
+/// [1]: https://eips.ethereum.org/EIPS/eip-55
+pub fn encode(input: &(impl AsRef<[u8]> + ?Sized)) -> String {
```

Comment:
```suggestion
pub fn encode(input: T) -> String 
where
       T: AsRef<[u8]> + ?Sized,
{
```

I think we may only need the reference for `encode_iter` since it has that implicit lifetime...

Reply:
Fixed in [e93dc37](https://github.com/nilz3ro/casper-node/pull/5/commits/e93dc3787a9983c352ebfdba0fc6bccb8f4d7c41).